pkg/kubelet/gpu/nvidia/nvidia.go

```diff
@@ -0,0 +1,171 @@
+	gputypes "k8s.io/kubernetes/pkg/kubelet/gpu/types"
+)
+
+var (
+	gpuMutex sync.Mutex
+)
+
+// TODO: If use NVML in the future, the implementation could be more complex,
+// and more powerful!
+
+type NvidiaGPU struct {
+	gpuInfo []gputypes.GPUDevice
+}
+
+func ProbePlugin() (gputypes.GPUPlugin, error) {
+	var nvidiaGPU NvidiaGPU
+
+	err := nvidiaGPU.InitPlugin()
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &nvidiaGPU, nil
+}
+
+// Get the vendor information.
+func (nvidiaGPU *NvidiaGPU) Vendor() string {
+	return gpuutil.Vendor
+}
+
+func (nvidiaGPU *NvidiaGPU) InitPlugin() error {
+	gpuMutex.Lock()
+	defer gpuMutex.Unlock()
+
+	if err := gpuutil.Init(); err != nil {
+		return err
+	}
+
+	allPaths, err := gpuutil.Discovery()
+
+	if err != nil {
+		return err
+	}
+
+	// Initialize the Nvidia device information.
+	for _, path := range allPaths {
+		nvidiaGPU.gpuInfo = append(nvidiaGPU.gpuInfo, gputypes.GPUDevice{Path: path, Status: gputypes.GPUFree, ContainerID: ""})
+	}
+
+	return nil
+}
+
+// Release the Plugin, could be useful once we use NVML.
+func (nvidiaGPU *NvidiaGPU) ReleasePlugin() error {
+	gpuMutex.Lock()
+	defer gpuMutex.Unlock()
+
+	nvidiaGPU.gpuInfo = nil
+
+	return nil
+}
+
+func (nvidiaGPU *NvidiaGPU) Capacity() (int, error) {
+	gpuMutex.Lock()
+	defer gpuMutex.Unlock()
+
+	return len(nvidiaGPU.gpuInfo), nil
+}
+
+func (nvidiaGPU *NvidiaGPU) AvailableGPUs() int {
+	gpuMutex.Lock()
+	defer gpuMutex.Unlock()
+
+	freeGPUs := 0
+	for _, gpuItem := range nvidiaGPU.gpuInfo {
+		if gpuItem.Status == gputypes.GPUFree {
+			freeGPUs += 1
+		}
+	}
+
+	return freeGPUs
+}
+
+func (nvidiaGPU *NvidiaGPU) AllocateGPU(number int) (allocPaths []string) {
+	availableGPUs := nvidiaGPU.AvailableGPUs()
+
+	gpuMutex.Lock()
+	defer gpuMutex.Unlock()
+
+	if availableGPUs < number {
+		return
+	}
+
+	allocatedGPU := 0
+
+	for _, gpuItem := range nvidiaGPU.gpuInfo {
+		if allocatedGPU >= number {
+			return
+		}
+
+		if gpuItem.Status == gputypes.GPUFree {
+			allocPaths = append(allocPaths, gpuItem.Path)
+			gpuItem.Status = gputypes.GPUInUse
+
+			allocatedGPU++
+		}
+	}
+
+	return
+}
+
+func (nvidiaGPU *NvidiaGPU) UpdateContainerID(containerID string, Paths []string) {
+	gpuMutex.Lock()
+	defer gpuMutex.Unlock()
+
+	for _, path := range Paths {
+		for _, gpuItem := range nvidiaGPU.gpuInfo {
+			if path == gpuItem.Path {
+				gpuItem.ContainerID = containerID
+			}
+		}
+	}
+}
+
+func (nvidiaGPU *NvidiaGPU) FreeGPUByContainer(containerID string) {
+	gpuMutex.Lock()
+	defer gpuMutex.Unlock()
+
+	for _, gpuItem := range nvidiaGPU.gpuInfo {
+		if gpuItem.ContainerID == containerID {
+			gpuItem.Status = gputypes.GPUFree
+		}
+	}
+}
+
+func (nvidiaGPU *NvidiaGPU) FreeGPUByPaths(paths []string) {
+	gpuMutex.Lock()
+	defer gpuMutex.Unlock()
+
+	for _, path := range paths {
+		for _, gpuItem := range nvidiaGPU.gpuInfo {
+			if gpuItem.Path == path {
+				gpuItem.Status = gputypes.GPUFree
+				continue
```

Comment:
This continue does nothing since there's no more code in the inner for loop anyways.